File: harness/coppice/internal/web/static/_tests/held.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import { reset, element } from './browser-stub.mjs';
import { render, paneRows } from '../roster.js';
import { titleFor, rowFromEvent } from '../floor.js';
import { needsYou, heldLine } from '../pane.js';
import { phoneQueue } from '../dock.js';

// A held ask is one a foreman hears first. It waits under working as
// waiting on the task's foreman, and it is not the operator's until the
// hold ends.

const HELD = { by: 'w1:p9', task: 't1', task_label: 'review-team', since: 40, until: 160 };

function p(id, state, held) {
  return {
    id, label: id, state, harness: 'claude', ts: 100,
    ask: state === 'blocked' ? { id: 'ask-' + id, summary: 'rm -rf build/', tier: 'undoable' } : null,
    held,
  };
}

const PANES = [p('a', 'blocked', HELD), p('b', 'blocked'), p('c', 'working')];

test('needsYou is a blocked pane whose ask no foreman holds', () => {
  assert.equal(needsYou(PANES[0]), false);
  assert.equal(needsYou(PANES[1]), true);
  assert.equal(needsYou(PANES[2]), false);
  assert.equal(needsYou(null), false);
});

test('the waiting line names the task and how long the ask has waited', () => {
  assert.equal(heldLine(HELD, 100), "waiting on review-team's foreman · 1m");
  assert.equal(heldLine({ task: 't7' }, 100), "waiting on t7's foreman");
});

test('a held pane is a working row with its waiting line, never an ask card', () => {
  reset();
  const root = element('screen-roster');
  root.append(element('queue-head'), element('roster'), element('roster-empty'));
  element('queue-head').append(element('queue-title'), element('queue-sub'));
  const rows = paneRows(PANES, 100);
  assert.equal(rows[0].state, 'working');
  assert.equal(rows[0].held, "waiting on review-team's foreman · 1m");
  render(rows);
  assert.equal(element('queue-title').textContent, '1 needs you');
  const cards = root.querySelectorAll('.ask').map((li) => li.dataset.row);
  assert.deepEqual(cards, ['b']);
  const held = root.querySelector('[data-row="a"]');
  assert.match(held.textContent, /waiting on review-team's foreman/);
});

test('the tab title and the phone queue count only what needs the operator', () => {
  assert.equal(titleFor(PANES), '1 · coppice');
  assert.deepEqual(phoneQueue(paneRows(PANES, 100)).map((r) => r.id), ['b']);
});

test('a state event carries the hold, and one with no hold ends it', () => {
  const held = rowFromEvent(PANES[1], { event: 'state', pane: 'b', state: 'blocked', ask: PANES[1].ask, held: HELD });
  assert.deepEqual(held.held, HELD);
  const free = rowFromEvent(held, { event: 'state', pane: 'b', state: 'blocked', ask: PANES[1].ask });
  assert.equal(free.held, null);
  assert.equal(needsYou(free), true);
});

File: harness/coppice/internal/tui/held_test.go
package tui

import (
	"strings"
	"testing"
)

// heldList is a pane.list with one ask a foreman holds and one it does
// not.
func heldList() []map[string]any {
	return []map[string]any{
		{"id": "w1:p1", "label": "writer", "state": "blocked", "ts": 100.0, "task": "t2",
			"ask":  map[string]any{"id": "ask-1", "summary": "rm -rf build/", "tier": "undoable"},
			"held": map[string]any{"by": "w1:p9", "task": "t1", "task_label": "review-team", "since": 40.0, "until": 160.0}},
		{"id": "w1:p2", "label": "loner", "state": "blocked", "ts": 100.0,
			"ask": map[string]any{"id": "ask-2", "summary": "git push", "tier": "permanent"}},
	}
}

func TestAHeldAskRendersUnderWorkingAndNotUnderNeedsYou(t *testing.T) {
	m := &Model{}
	m.Apply(heldList(), 100)
	if m.NeedYou != 1 {
		t.Fatalf("NeedYou = %d, want 1: a held ask is not the operator's yet", m.NeedYou)
	}
	secs := Group(m.Rows)
	if len(secs[0].Rows) != 1 || secs[0].Rows[0].ID != "w1:p2" {
		t.Fatalf("NEEDS YOU = %+v", secs[0].Rows)
	}
	if len(secs[1].Rows) != 1 || secs[1].Rows[0].ID != "w1:p1" {
		t.Fatalf("WORKING = %+v", secs[1].Rows)
	}
	text := stripSGR(strings.Join(Render(m, 100, 20), "\n"))
	if !strings.Contains(text, "waiting on review-team's foreman · 1m") {
		t.Fatalf("no waiting line:\n%s", text)
	}
	if !strings.Contains(text, "▾ t2  1 working") || !strings.Contains(text, "▾ no directory  1 needs") {
		t.Fatalf("the held row does not count as working in its group:\n%s", text)
	}
}

// A state event with no held field ends the hold on the row: that is how
// the ask reaches NEEDS YOU at the deadline.
func TestAnEventWithNoHoldMovesTheAskToNeedsYou(t *testing.T) {
	m := &Model{}
	m.Apply(heldList(), 100)
	m.Event(map[string]any{"pane": "w1:p1", "state": "blocked", "source": "gate",
		"ask": map[string]any{"id": "ask-1", "summary": "rm -rf build/"}})
	if m.NeedYou != 2 {
		t.Fatalf("NeedYou = %d, want 2", m.NeedYou)
	}
	m.Event(map[string]any{"pane": "w1:p1", "state": "blocked", "source": "gate",
		"ask":  map[string]any{"id": "ask-1", "summary": "rm -rf build/"},
		"held": map[string]any{"by": "w1:p9", "task": "t1", "task_label": "review-team", "since": 40.0}})
	if m.NeedYou != 1 {
		t.Fatalf("a held event counted: NeedYou = %d", m.NeedYou)
	}
}
